File: skills/sci-plot/assets/case-packs/rf-0063/clinical_heatmap.py
from collections import Counter
from pathlib import Path
import csv
import math

import matplotlib

matplotlib.use("Agg")

import matplotlib.pyplot as plt
from matplotlib.cm import ScalarMappable
from matplotlib.colors import LinearSegmentedColormap, Normalize, to_rgba
from matplotlib.patches import Patch
import numpy as np


BASE_DIR = Path(__file__).resolve().parent
DATA_PATH = BASE_DIR / "clinical_heatmap_data.tsv"
OUTPUT_PATH = BASE_DIR / "clinical_heatmap.png"


matplotlib.rcParams.update(
    {
        "font.family": "sans-serif",
        "font.sans-serif": ["Arial", "Helvetica", "DejaVu Sans"],
        "axes.unicode_minus": False,
        "figure.facecolor": "white",
        "savefig.facecolor": "white",
    }
)


def number(value):
    if value is None or str(value).strip() in {"", "NA", "NaN"}:
        return None
    result = float(value)
    return result if math.isfinite(result) else None


def load_rows():
    with DATA_PATH.open("r", encoding="utf-8-sig", newline="") as handle:
        reader = csv.DictReader(handle, delimiter="\t")
        rows = list(reader)
    required = {
        "patient_id",
        "mrd_status",
        "mrd_ctDNA_ppm_log",
        "mol_ml_peudolog",
        "cfdna_input_ng",
        "main_histology",
        "Stage",
        "lesion1_size_pathology",
        "pack_years_truncated",
        "luad_subtype",
        "oncogenic_event",
    }
    missing = required.difference(reader.fieldnames or [])
    if missing:
        raise ValueError(f"Missing required columns: {', '.join(sorted(missing))}")
    if len(rows) != 171 or len({row["patient_id"] for row in rows}) != 171:
        raise ValueError("Expected 171 unique patients")
    for row in rows:
        row["oncogenic_event"] = "None" if row["oncogenic_event"] in {"", "NA"} else row["oncogenic_event"]
        row["luad_subtype"] = "NA" if row["luad_subtype"] in {"", "NA"} else row["luad_subtype"].title()
    rows.sort(key=lambda row: (0 if row["main_histology"] == "LUAD" else 1, number(row["mrd_ctDNA_ppm_log"]) or 0.0))
    return rows


def numeric_array(rows, key):
    return np.asarray([np.nan if number(row[key]) is None else number(row[key]) for row in rows], dtype=float)


def colors_for_categories(rows, key, palette, fallback="#D9D9D9"):
    return np.asarray([to_rgba(palette.get(row[key], fallback)) for row in rows], dtype=float)


def colors_for_values(values, cmap, minimum, maximum, missing="#D9D9D9"):
    norm = Normalize(minimum, maximum, clip=True)
    result = []
    for value in values:
        result.append(to_rgba(missing) if not np.isfinite(value) else cmap(norm(value)))
    return np.asarray(result, dtype=float)


def style_bar_axis(axis, ticks, labels):
    axis.set_yticks(ticks)
    axis.set_yticklabels(labels, fontsize=7.5, color="#38434C")
    axis.tick_params(axis="y", width=0.7, length=3, pad=2)
    axis.tick_params(axis="x", bottom=False, labelbottom=False)
    axis.spines["top"].set_visible(False)
    axis.spines["right"].set_visible(False)
    axis.spines["left"].set_color("#7E878E")
    axis.spines["bottom"].set_color("#7E878E")
    axis.grid(axis="y", color="#E8EBED", linewidth=0.65, zorder=0)
    axis.set_axisbelow(True)


def draw_track(figure, bounds, colors):
    axis = figure.add_axes(bounds)
    axis.imshow(colors[np.newaxis, :, :], aspect="auto", interpolation="nearest")
    axis.set_xticks([])
    axis.set_yticks([])
    for spine in axis.spines.values():
        spine.set_visible(False)
    return axis


def add_legend(figure, handles, title, anchor, columns=1):
    legend = figure.legend(
        handles=handles,
        loc="lower left",
        bbox_to_anchor=anchor,
        ncol=columns,
        frameon=False,
        fontsize=8.5,
        title=title,
        title_fontsize=9.5,
        handlelength=1.0,
        handleheight=0.9,
        handletextpad=0.45,
        columnspacing=0.9,
        borderaxespad=0,
        labelspacing=0.35,
    )
    legend._legend_box.align = "left"


def add_colorbar(figure, cmap, limits, ticks, label, bounds):
    axis = figure.add_axes(bounds)
    colorbar = figure.colorbar(ScalarMappable(norm=Normalize(*limits), cmap=cmap), cax=axis, orientation="horizontal")
    colorbar.set_ticks(ticks)
    colorbar.ax.tick_params(labelsize=8.0, length=2.0, pad=1.5, colors="#38434C")
    colorbar.outline.set_visible(False)
    figure.text(bounds[0], bounds[1] + bounds[3] + 0.008, label, fontsize=9.3, fontweight="bold", color="#28323A", ha="left")


def draw_heatmap(rows):
    count = len(rows)
    x = np.arange(count)
    left = 0.058
    width = 0.738
    right_label = left + width + 0.010
    histology_palette = {"LUAD": "#D0BB7E", "Non-LUAD": "#91C7BA"}
    stage_palette = {"I": "#E8EAF5", "II": "#7AA8CE", "III": "#345C9F"}
    subtype_palette = {
        "Invasive Mucinous": "#EAB874",
        "Lepidic": "#A6C7DB",
        "Papillary": "#4B7BB1",
        "Acinar": "#B3D291",
        "Cribriform": "#6CA956",
        "Micropapillary": "#DE9493",
        "Solid": "#CA2E2E",
    }
    event_palette = {"None": "#E5E5E5", "EGFR mutated": "#602A84", "MET exon 14 skipped": "#CCAC68"}
    ctdna_palette = {"Detected": "#8FD0F3", "Not Detected": "#BFC2C5"}
    size_cmap = LinearSegmentedColormap.from_list("tumor_size", ["#F7F8FC", "#A5A9C8", "#2F386E"])
    smoking_cmap = LinearSegmentedColormap.from_list("smoking", ["#F7F7F7", "#A7A7A7", "#171717"])
    figure = plt.figure(figsize=(20, 11.5))
    figure.text(left, 0.958, "Clinicopathological landscape of early-stage lung cancer", fontsize=22, fontweight="bold", color="#20272D", ha="left")
    figure.text(left, 0.925, "171 patients ordered by histology and preoperative ctDNA abundance", fontsize=10.8, color="#64717B", ha="left")

    ctdna_values = numeric_array(rows, "mrd_ctDNA_ppm_log")
    molecule_values = numeric_array(rows, "mol_ml_peudolog")
    input_values = numeric_array(rows, "cfdna_input_ng")
    axis_ctdna = figure.add_axes([left, 0.735, width, 0.145])
    axis_ctdna.bar(x, ctdna_values, width=0.82, color=[ctdna_palette[row["mrd_status"]] for row in rows], edgecolor="none", zorder=2)
    axis_ctdna.set_xlim(-0.6, count - 0.4)
    axis_ctdna.set_ylim(0, 5.65)
    style_bar_axis(axis_ctdna, [0, 1, 2, 3, 4, 5], ["0", "10", "100", "1,000", "10,000", "100,000"])

    axis_molecule = figure.add_axes([left, 0.655, width, 0.072], sharex=axis_ctdna)
    axis_molecule.bar(x, molecule_values, width=0.82, color="#3C5488", edgecolor="none", zorder=2)
    axis_molecule.set_ylim(0, 5.9)
    style_bar_axis(axis_molecule, [0, 1, 3, 5], ["0", "10", "1,000", "100,000"])

    histology_colors = colors_for_categories(rows, "main_histology", histology_palette)
    stage_colors = colors_for_categories(rows, "Stage", stage_palette)
    size_values = numeric_array(rows, "lesion1_size_pathology")
    smoking_values = numeric_array(rows, "pack_years_truncated")
    subtype_colors = colors_for_categories(rows, "luad_subtype", subtype_palette)
    event_colors = colors_for_categories(rows, "oncogenic_event", event_palette)
    tracks = [
        (0.616, histology_colors, "Histology"),
        (0.577, stage_colors, "pTNM stage"),
        (0.538, colors_for_values(size_values, size_cmap, 5, 120), "Tumor size (mm)"),
        (0.499, colors_for_values(smoking_values, smoking_cmap, 0, 136), "Smoking (pack years)"),
        (0.460, subtype_colors, "LUAD subtype"),
        (0.421, event_colors, "Oncogenic event"),
    ]
    for bottom, colors, label in tracks:
        draw_track(figure, [left, bottom, width, 0.032], colors)
        figure.text(right_label, bottom + 0.016, label, fontsize=9.7, color="#2A333A", va="center", ha="left")

    axis_input = figure.add_axes([left, 0.337, width, 0.076], sharex=axis_ctdna)
    axis_input.bar(x, input_values, width=0.78, color="#B09C85", edgecolor="none", zorder=2)
    axis_input.set_ylim(0, 52)
    style_bar_axis(axis_input, [0, 20, 40], ["0", "20", "40"])
    figure.text(right_label, 0.375, "Input cfDNA (ng)", fontsize=9.7, color="#2A333A", va="center", ha="left")

    axis_ids = figure.add_axes([left, 0.292, width, 0.038], sharex=axis_ctdna)
    axis_ids.set_xlim(-0.6, count - 0.4)
    axis_ids.set_ylim(0, 1)
    axis_ids.set_xticks(x)
    axis_ids.set_xticklabels([row["patient_id"].replace("CRUK", "") for row in rows], rotation=90, fontsize=5.0, color="#364047")
    axis_ids.tick_params(axis="x", length=0, pad=1)
    axis_ids.set_yticks([])
    for spine in axis_ids.spines.values():
        spine.set_visible(False)
    figure.text(right_label, 0.311, "CRUK ID", fontsize=9.7, color="#2A333A", va="center", ha="left")

    split = sum(row["main_histology"] == "LUAD" for row in rows) - 0.5
    for axis in [axis_ctdna, axis_molecule, axis_input]:
        axis.axvline(split, color="white", linewidth=2.2, zorder=5)
    for bottom, _, _ in tracks:
        separator = figure.add_axes([left + width * (split + 0.5) / count - 0.0006, bottom, 0.0012, 0.032])
        separator.set_facecolor("white")
        separator.set_xticks([])
        separator.set_yticks([])
        for spine in separator.spines.values():
            spine.set_visible(False)

    figure.text(right_label, 0.807, "ctDNA (PPM)", fontsize=10.0, fontweight="bold", color="#2A333A", va="center", ha="left")
    figure.text(right_label, 0.691, "Tumor molecules/mL", fontsize=10.0, fontweight="bold", color="#2A333A", va="center", ha="left")

    add_legend(figure, [Patch(color=color, label=label) for label, color in histology_palette.items()], "Histology", (0.058, 0.155))
    add_legend(figure, [Patch(color=color, label=label) for label, color in ctdna_palette.items()], "ctDNA", (0.150, 0.155))
    add_legend(figure, [Patch(color=color, label=label) for label, color in stage_palette.items()], "pTNM stage", (0.260, 0.155))
    add_legend(figure, [Patch(color=color, label=label) for label, color in subtype_palette.items()], "LUAD subtype", (0.365, 0.125), columns=2)
    add_legend(figure, [Patch(color=color, label=label) for label, color in event_palette.items()], "Oncogenic event", (0.615, 0.145))
    add_colorbar(figure, size_cmap, (5, 120), [5, 50, 100, 120], "Tumor size (mm)", [0.058, 0.055, 0.145, 0.018])
    add_colorbar(figure, smoking_cmap, (0, 136), [0, 50, 100, 136], "Smoking (pack years)", [0.245, 0.055, 0.145, 0.018])
    figure.text(0.955, 0.055, "Bars and tracks share patient order", fontsize=7.6, color="#7A848C", ha="right")
    figure.savefig(OUTPUT_PATH, dpi=240)
    plt.close(figure)


def main():
    rows = load_rows()
    if Counter(row["main_histology"] for row in rows) != Counter({"LUAD": 94, "Non-LUAD": 77}):
        raise ValueError("Unexpected histology counts")
    draw_heatmap(rows)


if __name__ == "__main__":
    main()

File: skills/sci-plot/assets/case-packs/rf-0063/clinical_heatmap.R
resolve_script_dir <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  file_args <- grep("^--file=", args, value = TRUE)
  if (length(file_args)) return(dirname(normalizePath(sub("^--file=", "", file_args[1]), winslash = "/", mustWork = FALSE)))
  frames <- sys.frames()
  if (length(frames)) for (i in rev(seq_along(frames))) if (!is.null(frames[[i]]$ofile)) return(dirname(normalizePath(frames[[i]]$ofile, winslash = "/", mustWork = FALSE)))
  normalizePath(getwd(), winslash = "/", mustWork = FALSE)
}

script_dir <- resolve_script_dir()
data_path <- file.path(script_dir,"clinical_heatmap_data.tsv")
output_path <- file.path(script_dir,"clinical_heatmap_r.png")
if (!file.exists(data_path)) stop("Data file not found")

dat <- read.delim(data_path,sep="\t",stringsAsFactors=FALSE,check.names=FALSE,fileEncoding="UTF-8-BOM",na.strings=c("","NA","NaN"))
required <- c("patient_id","mrd_status","mrd_ctDNA_ppm_log","mol_ml_peudolog","cfdna_input_ng","main_histology","Stage","lesion1_size_pathology","pack_years_truncated","luad_subtype","oncogenic_event")
if (!all(required %in% names(dat)) || nrow(dat)!=171 || length(unique(dat$patient_id))!=171) stop("Unexpected clinical data")
dat$oncogenic_event[is.na(dat$oncogenic_event)|!nzchar(dat$oncogenic_event)] <- "None"
dat$luad_subtype <- ifelse(is.na(dat$luad_subtype)|!nzchar(dat$luad_subtype),"NA",tools::toTitleCase(tolower(dat$luad_subtype)))
dat <- dat[order(ifelse(dat$main_histology=="LUAD",0,1),ifelse(is.finite(dat$mrd_ctDNA_ppm_log),dat$mrd_ctDNA_ppm_log,0)),]
if (!identical(as.integer(table(factor(dat$main_histology,levels=c("LUAD","Non-LUAD")))),c(94L,77L))) stop("Unexpected histology counts")

histology_palette <- c(LUAD="#D0BB7E","Non-LUAD"="#91C7BA")
stage_palette <- c(I="#E8EAF5",II="#7AA8CE",III="#345C9F")
subtype_palette <- c("Invasive Mucinous"="#EAB874",Lepidic="#A6C7DB",Papillary="#4B7BB1",Acinar="#B3D291",Cribriform="#6CA956",Micropapillary="#DE9493",Solid="#CA2E2E")
event_palette <- c(None="#E5E5E5","EGFR mutated"="#602A84","MET exon 14 skipped"="#CCAC68")
ctdna_palette <- c(Detected="#8FD0F3","Not Detected"="#BFC2C5")
size_palette <- colorRampPalette(c("#F7F8FC","#A5A9C8","#2F386E"))(256)
smoking_palette <- colorRampPalette(c("#F7F7F7","#A7A7A7","#171717"))(256)

value_colours <- function(values,limits,palette,missing="#D9D9D9") {
  index <- round(pmax(0,pmin(1,(values-limits[1])/diff(limits)))*(length(palette)-1))+1
  result <- palette[index]
  result[!is.finite(values)] <- missing
  result
}

category_colours <- function(values,palette,missing="#D9D9D9") {
  result <- unname(palette[values])
  result[is.na(result)] <- missing
  result
}

x0 <- 0.058
x1 <- 0.796
label_x <- x1+0.010
n <- nrow(dat)
cell_width <- (x1-x0)/n
x_left <- x0+(seq_len(n)-1)*cell_width
x_right <- x_left+cell_width*0.88

draw_bar <- function(values,bottom,top,maximum,ticks,tick_labels,colours,label) {
  for(i in seq_along(ticks)) {
    y <- bottom+(ticks[i]/maximum)*(top-bottom)
    segments(x0,y,x1,y,col="#E8EBED",lwd=0.8)
    text(x0-0.006,y,tick_labels[i],adj=c(1,0.5),cex=0.62,col="#38434C")
  }
  for(i in seq_len(n)) if(is.finite(values[i])) rect(x_left[i],bottom,x_right[i],bottom+pmin(values[i],maximum)/maximum*(top-bottom),col=colours[i],border=NA)
  segments(x0,bottom,x1,bottom,col="#7E878E",lwd=0.8)
  segments(x0,bottom,x0,top,col="#7E878E",lwd=0.8)
  text(label_x,(bottom+top)/2,label,adj=c(0,0.5),cex=0.82,font=2,col="#2A333A")
}

draw_track <- function(colours,bottom,label) {
  top <- bottom+0.032
  for(i in seq_len(n)) rect(x0+(i-1)*cell_width,bottom,x0+i*cell_width,top,col=colours[i],border=NA)
  text(label_x,(bottom+top)/2,label,adj=c(0,0.5),cex=0.79,col="#2A333A")
}

draw_legend <- function(x,y,title,labels,colours,ncol=1,width=0.09) {
  text(x,y+0.035,title,adj=c(0,0),cex=0.78,font=2,col="#28323A")
  for(i in seq_along(labels)) {
    column <- (i-1)%%ncol
    row <- (i-1)%/%ncol
    xx <- x+column*width
    yy <- y-row*0.025
    rect(xx,yy-0.007,xx+0.010,yy+0.007,col=colours[i],border=NA)
    text(xx+0.014,yy,labels[i],adj=c(0,0.5),cex=0.64,col="#3B464E")
  }
}

draw_gradient <- function(x,y,w,palette,ticks,limits,title) {
  text(x,y+0.035,title,adj=c(0,0),cex=0.76,font=2,col="#28323A")
  for(i in seq_along(palette)) rect(x+(i-1)/length(palette)*w,y,x+i/length(palette)*w,y+0.018,col=palette[i],border=NA)
  for(v in ticks) text(x+(v-limits[1])/diff(limits)*w,y-0.006,v,adj=c(0.5,1),cex=0.60,col="#38434C")
}

png(output_path,width=6000,height=3450,res=300,bg="white",type=if(.Platform$OS.type=="windows") "windows" else "cairo")
par(family="sans",mar=c(0,0,0,0),xpd=NA)
plot.new()
plot.window(xlim=c(0,1),ylim=c(0,1))
text(x0,0.958,"Clinicopathological landscape of early-stage lung cancer",adj=c(0,1),cex=1.95,font=2,col="#20272D")
text(x0,0.918,"171 patients ordered by histology and preoperative ctDNA abundance",adj=c(0,1),cex=0.96,col="#64717B")

ctdna_colours <- category_colours(dat$mrd_status,ctdna_palette)
draw_bar(as.numeric(dat$mrd_ctDNA_ppm_log),0.735,0.880,5.65,0:5,c("0","10","100","1,000","10,000","100,000"),ctdna_colours,"ctDNA (PPM)")
draw_bar(as.numeric(dat$mol_ml_peudolog),0.655,0.727,5.9,c(0,1,3,5),c("0","10","1,000","100,000"),rep("#3C5488",n),"Tumor molecules/mL")
draw_track(category_colours(dat$main_histology,histology_palette),0.616,"Histology")
draw_track(category_colours(dat$Stage,stage_palette),0.577,"pTNM stage")
draw_track(value_colours(as.numeric(dat$lesion1_size_pathology),c(5,120),size_palette),0.538,"Tumor size (mm)")
draw_track(value_colours(as.numeric(dat$pack_years_truncated),c(0,136),smoking_palette),0.499,"Smoking (pack years)")
draw_track(category_colours(dat$luad_subtype,subtype_palette),0.460,"LUAD subtype")
draw_track(category_colours(dat$oncogenic_event,event_palette),0.421,"Oncogenic event")
draw_bar(as.numeric(dat$cfdna_input_ng),0.337,0.413,52,c(0,20,40),c("0","20","40"),rep("#B09C85",n),"Input cfDNA (ng)")

split_x <- x0+94*cell_width
segments(split_x,0.337,split_x,0.880,col="white",lwd=2.5)
for(i in seq_len(n)) text(x0+(i-0.5)*cell_width,0.323,gsub("CRUK","",dat$patient_id[i]),srt=90,adj=c(1,0.5),cex=0.39,col="#364047")
text(label_x,0.311,"CRUK ID",adj=c(0,0.5),cex=0.79,col="#2A333A")

draw_legend(0.058,0.165,"Histology",names(histology_palette),histology_palette)
draw_legend(0.150,0.165,"ctDNA",names(ctdna_palette),ctdna_palette)
draw_legend(0.260,0.165,"pTNM stage",names(stage_palette),stage_palette)
draw_legend(0.365,0.165,"LUAD subtype",names(subtype_palette),subtype_palette,2,0.105)
draw_legend(0.615,0.165,"Oncogenic event",names(event_palette),event_palette)
draw_gradient(0.058,0.055,0.145,size_palette,c(5,50,100,120),c(5,120),"Tumor size (mm)")
draw_gradient(0.245,0.055,0.145,smoking_palette,c(0,50,100,136),c(0,136),"Smoking (pack years)")
text(0.955,0.055,"Bars and tracks share patient order",adj=c(1,0),cex=0.64,col="#7A848C")
dev.off()
cat("Saved:",output_path,"\n")
